ksql-engine/src/main/java/io/confluent/ksql/codegen/FunctionArguments.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.ksql.codegen;
+
+import org.apache.kafka.connect.data.Schema;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Stack;
+
+public class FunctionArguments {
+  private Stack<List<Schema.Type>> args = new Stack<>();
+
+  public void beginFunction() {
+    args.push(new ArrayList<>());
+  }
+
+  public void addArgumentType(final Schema.Type type) {
+    if (!args.isEmpty()) {
```

Comment:
Should we throw an exception if args is empty? This would happen if `addArgumentType` was called before `beginFunction`, which seems like an error.

I am also curious about what function the stack serves, since it never seems to grow beyond a size of 1.

Reply:
it does indeed grow beyond 1 if you have nested udf calls.
if `args.isEmpty()` we aren't in a function call so there is nothing to do. It isn't an error condition

Reply:
I wouldn't it be a bug to call addArgumentType from outside a function call?

Reply:
Oh I see its just called by default for any expression in case it happens to be a function arg.

Reply:
correct - it is so that the logic for checking etc is in one place rather than spread over 3 different classes